migration/ubuntu/20.04-to-22.04-openstack.md

```diff
@@ -0,0 +1,256 @@
+> If the environment depend on these IPs in any way, please communicate this well in advance.
+
+## Prepare
+
+These steps can be done without any disruption to the running cluster.
+
+- Prepare environment variables
+
+  ```bash
+  export CK8S_CONFIG_PATH="/path/to/cluster/config"
+  export CK8S_KUBESPRAY_REPOSITORY_PATH="/path/to/compliantkubernetes-kubespray"
+  export CK8S_APPS_REPOSITORY_PATH="/path/to/compliantkubernetes-apps"
+  export KUBECONFIG="${CK8S_CONFIG_PATH}/.state/kube_config_<sc|wc>.yaml"
+  ```
+
+- In `${CK8S_CONFIG_PATH}/<sc|wc>-config/cluster.tfvars` update the following:
+
+  - Change global `image_uuid` to the ID of Ubuntu 24.04
+  - Configure `image_id` for all existing nodes to the old image id
+
+- Verify by running `terraform plan` and make sure there's no changes on any on the existing nodes
+- Add a new control plane node that doesn't specify `image_id` (To use the new image)
+- Verify by running `terraform plan` and make sure there's a new node added with the correct image
+- Add the rest of the new control plane nodes
+- Run `terraform apply` to add all the new nodes
+- Backup the current state:
+
+  ```bash
+  pushd "${CK8S_CONFIG_PATH}"
+  cp -r <sc|wc>-config <sc|wc>-config-backup
+  popd
+  ```
+
+- Generate new inventory:
+
+  ```bash
+  ./gen-inventory.sh <sc|wc>
+  ```
+
+- Reorder inventory so the old control plane nodes are first in each category they occur in
+- Add the new control plane floating IPs into group vars under the key `supplementary_addresses_in_ssl_keys`
+
+> [!NOTE]
+> Adding node IPs to `supplementary_addresses_in_ssl_keys` is not needed if you are using a loadbalancer in front of the API server or only use DNS records.
+> You can check to see what `supplementary_addresses_in_ssl_keys` is currently pointing to, and if it's not the IPs of the control plane nodes then this can be left alone.
+
+## Execute
+
+These steps will cause disruptions in the cluster.
+
+### Replace control plane nodes
+
+- Add new control plane nodes using Kubespray:
+
+  ```bash
+  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
+  ck8s-kubespray apply <sc|wc> --limit=etcd,kube_control_plane -b -e=ignore_assert_errors=true --skip-tags=multus
+  popd
+  ```
+
+- Update control plane proxy using Kubespray:
+
+  ```bash
+  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
+  ck8s-kubespray apply <sc|wc> --tags=nginx -b
+  popd
+  ```
+
+- Update the NetworkPolicies:
+
+  ```bash
+  pushd "${CK8S_APPS_REPOSITORY_PATH}"
+  ck8s update-ips <sc|wc> apply
+
+  ck8s ops helmfile <sc|wc> -l app=netpol -i apply
+  popd
+  ```
+
+- Reorder inventory so the new control plane nodes are first in each category they occur in
+- Set the public IP of the first new control plane node into group vars under the key `kube_oidc_apiserver_endpoint`
+
+> [!NOTE]
+> This is not needed if it's not pointing to an old control plane node directly.
+> Like if it's pointing to a loadbalancer or DNS or similar.
```

Comment:
```suggestion
> [!NOTE]
> Changing `kube_oidc_apiserver_endpoint` is not needed if it's not pointing to an old control plane node directly.
> Like if it's pointing to a loadbalancer or DNS or similar.
```

Just to avoid confusion, I think it would be nice if these !NOTEs were more specific in what they are referring to.

Same applies to the other !NOTEs in the document.

Reply:
https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/5feea8586cdaf1dab62242923751189edbce4293